File: BazaMobileHW/src/main/NotesApp/src/NotesRealization.kt
import kotlin.reflect.KClass

abstract class Note<Data : IdentifiableModel>(
    var data: Data,
    private val storage: Storage<Data>,
    private val dataClass: KClass<Data>
) : NoteProtocol {
    override val id: String
        get() = data.id

    init {
        storage.add(data)
    }

    override fun update(newData: IdentifiableModel) {
        if (dataClass.isInstance(newData)) {
            @Suppress("UNCHECKED_CAST")
            this.data = newData as Data
            storage.update(this.data)
        }
    }

    override fun willRemove() {
        storage.removeById(data.id)
    }

    override fun getAll(): List<IdentifiableModel> {
        return listOf(data)
    }

    override fun drawCanvas(): String {
        return "Базовая заметка ${data.id}"
    }
}

data class TextNoteModel(
    override val id: String,
    var name: String,
    var text: String
) : IdentifiableModel

data class ReminderNoteModel(
    override val id: String,
    var text: String,
    var isDone: Boolean
) : IdentifiableModel


class TextNote(
    data: TextNoteModel,
    storage: Storage<TextNoteModel>
) : Note<TextNoteModel>(data, storage, TextNoteModel::class) {
    override fun drawCanvas(): String {
        return "Текстовая заметка с именем ${data.name} содержит: ${data.text}"
    }
}

class ReminderNote(
    data: ReminderNoteModel,
    storage: Storage<ReminderNoteModel>
) : Note<ReminderNoteModel>(data, storage, ReminderNoteModel::class) {
    override fun drawCanvas(): String {
        return "Задача ${data.text} ${if (data.isDone) "выполнена" else "не выполнена"}"
    }
}

File: NotesApp/NotesRealization.swift
import Foundation

extension IdentifableModel {
    static func generateID() -> String {
        return UUID().uuidString
    }
}

protocol NoteProtocol : CanvasUnit{
    func willRemove()
    func update(_ data:IdentifableModel)
}

class Note<Data:IdentifableModel> : NoteProtocol {
    var id: String {return data.id}

    
    var data: Data{
        didSet{
            storage.update(data)
        }
    }
    private var storage : Storage<Data>
    
    init(data: Data, storage: Storage<Data>) {
        self.data = data
        self.storage = storage
        self.storage.add(data)
    }

    
    func update(_ newData:IdentifableModel) {
        if let typedData = newData as? Data {
            self.data = typedData
        }
    }
    
    func willRemove() {
        storage.remove(ById: data.id)
    }
    
    func drawCanvas() -> String {
        return "Базовая заметка \(data.id)"
    }
    func getAll() -> [any IdentifableModel] {
        return [data]
    }
}

struct TextNoteModel : IdentifableModel {
    var id: String
    var name:String
    var text: String
}

struct ReminderNoteModel : IdentifableModel {
    var id:String
    var text:String
    var isDone : Bool
    
}

class TextNote: Note<TextNoteModel> {
    override func drawCanvas() -> String {
        return "Текстовая заметка с именем \(data.name) содержит : \(data.text)"
    }
    
    
}

class ReminderNote: Note<ReminderNoteModel> {
    override func drawCanvas() -> String {
        return "Задача \(data.text)  \(data.isDone ? "выполнена" : "не выполнена")"
    }

}

